File: flutter_embed_unity_2022_3_android/android/src/main/kotlin/com/jamesncl/dev/flutter_embed_unity_android/messaging/SendToUnity.kt
package com.jamesncl.dev.flutter_embed_unity_android.messaging

import com.jamesncl.dev.flutter_embed_unity_android.constants.FlutterEmbedConstants.Companion.logTag
import com.jamesncl.dev.flutter_embed_unity_android.constants.FlutterEmbedConstants.Companion.methodNameOrientationChanged
import com.jamesncl.dev.flutter_embed_unity_android.constants.FlutterEmbedConstants.Companion.methodNamePauseUnity
import com.jamesncl.dev.flutter_embed_unity_android.constants.FlutterEmbedConstants.Companion.methodNameResumeUnity
import com.jamesncl.dev.flutter_embed_unity_android.constants.FlutterEmbedConstants.Companion.methodNameSendToUnity
import com.jamesncl.dev.flutter_embed_unity_android.unity.UnityPlayerSingleton
import com.unity3d.player.UnityPlayer
import io.flutter.Log
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel

class SendToUnity: MethodChannel.MethodCallHandler {

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        Log.i(logTag, call.method)

        when (call.method) {
            methodNameSendToUnity -> {
                val gameObjectMethodNameData = (call.arguments as List<*>).filterIsInstance<String>()
                UnityPlayer.UnitySendMessage(
                    gameObjectMethodNameData[0], // Unity game object name
                    gameObjectMethodNameData[1], // Game object method name
                    gameObjectMethodNameData[2]) // Data
            }
            methodNameOrientationChanged -> {
                UnityPlayerSingleton.getInstance()?.orientationChanged()
            }
            methodNamePauseUnity -> {
                UnityPlayerSingleton.getInstance()?.pause()
            }
            methodNameResumeUnity -> {
                UnityPlayerSingleton.getInstance()?.resume()
            }
            else -> {
                result.notImplemented()
            }
        }
    }
}

File: flutter_embed_unity_2022_3_ios/ios/Classes/Messaging/SendToUnity.swift
import Flutter

class SendToUnity {
    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
            switch call.method {
            case FlutterEmbedConstants.methodNameSendToUnity:
                let gameObjectMethodNameData = call.arguments as! [String]
                if(UnityPlayerSingleton.isInitialised) {
                    UnityPlayerSingleton.getInstance().sendMessageToGO(
                        withName: gameObjectMethodNameData[0],
                        functionName: gameObjectMethodNameData[1],
                        message: gameObjectMethodNameData[2])
                    result(true)
                }
                else {
                    debugPrint("Dropped message to Unity: Unity is not loaded yet")
                    result(false)
                }
            case FlutterEmbedConstants.methodNamePauseUnity:
                if(UnityPlayerSingleton.isInitialised) {
                    UnityPlayerSingleton.getInstance().pause(true)
                    result(true)
                }
                else {
                    debugPrint("Didn't pause Unity: Unity is not loaded yet")
                    result(false)
                }
            case FlutterEmbedConstants.methodNameResumeUnity:
                if(UnityPlayerSingleton.isInitialised) {
                    UnityPlayerSingleton.getInstance().pause(false)
                    result(true)
                }
                else {
                    debugPrint("Didn't resume Unity: Unity is not loaded yet")
                    result(false)
                }
            default:
              result(FlutterMethodNotImplemented)
            }
    }
}
